File: gradle/src/main/java/com/antwerkz/critter/CritterPlugin.java
package com.antwerkz.critter;

import java.io.File;
import java.util.concurrent.Callable;

import org.gradle.api.Action;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.artifacts.Configuration;
import org.gradle.api.plugins.JavaPlugin;
import static org.gradle.api.plugins.JavaPlugin.COMPILE_CONFIGURATION_NAME;
import org.gradle.api.plugins.JavaPluginConvention;
import org.gradle.api.tasks.SourceSet;

public class CritterPlugin implements Plugin<Project> {
  public static final String CRITTER_CONFIGURATION_NAME = "critter";

  @Override
  public void apply(final Project project) {
    project.getPlugins().apply(JavaPlugin.class);
    final CritterPluginExtension extension = project.getExtensions().create("critter", CritterPluginExtension.class);

    Configuration critterConfiguration = project.getConfigurations().create(CRITTER_CONFIGURATION_NAME).setVisible(false)
            .setTransitive(false).setDescription("The Critter libraries to be used for this project.");
    project.getConfigurations().getByName(COMPILE_CONFIGURATION_NAME).extendsFrom(critterConfiguration);
  
    project.getConvention().getPlugin(JavaPluginConvention.class).getSourceSets().all(new Action<SourceSet>() {
      public void execute(SourceSet sourceSet) {
        final String taskName = sourceSet.getTaskName("generate", "CritterSource");
        CritterTask task = project.getTasks().create(taskName, CritterTask.class);
        task.setDescription(String.format("Processes the %s java files for critter.",
            sourceSet.getName()));

        task.setSource(sourceSet.getAllJava().getFiles());
        task.setExtension(extension);

        task.getConventionMapping().map("critterClasspath",
            () -> project.getConfigurations().getByName(CRITTER_CONFIGURATION_NAME).copy().setTransitive(true));

        final String outputDirectoryName = String.format("%s/generated-src/critter/%s",
            project.getBuildDir(), sourceSet.getName());
        final File outputDirectory = new File(outputDirectoryName);
        task.setOutputDirectory(outputDirectory);
        sourceSet.getJava().srcDir(outputDirectory);

        project.getTasks().getByName(sourceSet.getCompileJavaTaskName()).dependsOn(taskName);
      }
    });
  }
}

File: gradle/src/main/kotlin/com/antwerkz/critter/CritterPlugin.kt
package dev.morphia.critter

import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.plugins.JavaPlugin
import org.gradle.api.plugins.JavaPlugin.COMPILE_CONFIGURATION_NAME
import org.gradle.api.plugins.JavaPluginConvention
import java.io.File

class CritterPlugin : Plugin<Project> {

    override fun apply(project: Project) {
        project.plugins.apply(JavaPlugin::class.java)
        val extension = project.extensions.create("critter", CritterPluginExtension::class.java)

        val critterConfiguration = project.configurations.create(CRITTER_CONFIGURATION_NAME).setVisible(false)
                .setTransitive(false).setDescription("The Critter libraries to be used for this project.")
        project.configurations.getByName(COMPILE_CONFIGURATION_NAME).extendsFrom(critterConfiguration)

        project.convention.getPlugin(JavaPluginConvention::class.java).sourceSets.all { sourceSet ->
            val taskName = sourceSet.getTaskName("generate", "CritterSource")
            val task = project.tasks.create(taskName, CritterTask::class.java)
            task.description = "Processes the ${sourceSet.name} java files for critter."

            task.setSource(sourceSet.allJava.files)
            task.extension = extension

            task.conventionMapping.map("critterClasspath"
            ) { project.configurations.getByName(CRITTER_CONFIGURATION_NAME).copy().setTransitive(true) }

            val outputDirectoryName = "${project.buildDir}/generated-src/critter/${sourceSet.name}"
            val outputDirectory = File(outputDirectoryName)
            task.outputDirectory = outputDirectory
            sourceSet.java.srcDir(outputDirectory)

            project.tasks.getByName(sourceSet.compileJavaTaskName).dependsOn(taskName)
        }
    }

    companion object {
        val CRITTER_CONFIGURATION_NAME = "critter"
    }
}
